afew/tests/test_mailmover.py

```diff
@@ -0,0 +1,162 @@
+
+def create_mail(msg, maildir, notmuch_db, tags, old=False):
+    email_message = email.message.EmailMessage()
+    # freezegun doesn't handle time zones properly when generating UNIX
+    # timestamps.  When the local timezone is UTC+2, the generated timestamp
+    # is 2 hours ahead of what it should be.  Due to this we need to make sure
+    # that the dates are always sufficiently far behind 2019-01-30 12:00 to
+    # handle up to UTC+12 .
+    if old:
+        email_message['Date'] = 'Wed, 10 Jan 2019 13:00:00 +0100'
+    else:
+        email_message['Date'] = 'Wed, 20 Jan 2019 13:00:00 +0100'
+    email_message['From'] = 'You <you@example.org>'
+    email_message['To'] = 'Me <me@example.com>'
+    email_message['Message-ID'] = make_msgid()
+    email_message.set_content(msg)
+
+    maildir_message = mailbox.MaildirMessage(email_message)
+    message_key = maildir.add(maildir_message)
+
+    fname = os.path.join(maildir._path, maildir._lookup(message_key))
+    notmuch_msg = notmuch_db.add_message(fname)
+    for tag in tags:
+        notmuch_msg.add_tag(tag, False)
+
+    # Remove the angle brackets automatically added around the message ID by make_msgid.
+    stripped_msgid = email_message['Message-ID'].strip('<>')
+    return (stripped_msgid, msg)
+
+
+@freeze_time("2019-01-30 12:00:00")
+class TestMailMover(unittest.TestCase):
+    def setUp(self):
+        self.test_dir = tempfile.mkdtemp()
+        os.environ['MAILDIR'] = self.test_dir
+
+        self.notmuch_config = os.path.join(self.test_dir, 'notmuch-config')
+        with open(self.notmuch_config, 'w+') as fp:
+            fp.write('[database]\npath={}\n\n[new]\ntags=new'.format(self.test_dir))
+        os.environ['NOTMUCH_CONFIG'] = self.notmuch_config
+        read_notmuch_settings()
+
+        # Create notmuch database
+        Database().open(create=True).close()
+
+        self.root = mailbox.Maildir(self.test_dir)
+        self.inbox = self.root.add_folder('inbox')
+        self.archive = self.root.add_folder('archive')
+        self.spam = self.root.add_folder('spam')
+
+        # Dict of rules that are passed to MailMover.
+        #
+        # The top level key represents a particular mail directory to work on.
+        #
+        # The second level key is the notmuch query that MailMover will execute,
+        # and its value is the directory to move the matching emails to.
+        self.rules = {
+            '.inbox': {
+                'tag:archive AND NOT tag:spam': '.archive',
+                'tag:spam': '.spam',
+            },
+            '.archive': {
+                'NOT tag:archive AND NOT tag:spam': '.inbox',
+                'tag:spam': '.spam',
+            },
+            '.spam': {
+                'NOT tag:spam AND tag:archive': '.archive',
+                'NOT tag:spam AND NOT tag:archive': '.inbox',
+            },
+        }
+
+
+    def tearDown(self):
+        shutil.rmtree(self.test_dir)
+
+
+    def assert_on_expected(self):
```

Comment:
I don't really like the way tests mutate `self.expect_*`, and we call that function afterwards. See for a alternative suggestion below.